api4/user.go

```diff
@@ -1129,6 +1156,61 @@ func updateUser(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(ruser.ToJson()))
 }
 
+// checkLoginProviderAttributes returns nil if the patch can be applied without overriding
+// attributes set by the user's login provider; otherwise, an error is returned whose text
+// is the name of the offending field.
+func checkLoginProviderAttributes(cfg *model.Config, ouser *model.User, patch *model.UserPatch) error {
+	LS := &cfg.LdapSettings
+	SS := &cfg.SamlSettings
+	// If LDAP or SAML is used, and either the first name or the last name are already set,
+	// or OAUTH is used, then neither the first name nor the last name may be modified
+	if ((patch.FirstName != nil && *patch.FirstName != ouser.FirstName) ||
+		(patch.LastName != nil && *patch.LastName != ouser.LastName)) &&
+		((ouser.IsLDAPUser() && (*LS.FirstNameAttribute != "" || *LS.LastNameAttribute != "")) ||
+			(ouser.IsSAMLUser() && (*SS.FirstNameAttribute != "" || *SS.LastNameAttribute != "")) ||
+			ouser.IsOAuthUser()) {
+		return errors.New("full name")
+	}
+	// If LDAP or SAML is used, and any of the following attributes have been specified
+	// in the settings, then their corresponding fields may not be changed
+	if ouser.IsLDAPUser() {
+		for _, attr := range []struct {
+			name       string
+			patchValue *string
+			userValue  *string
+			selector   *string
+		}{
+			{"nickname", patch.Nickname, &ouser.Nickname, LS.NicknameAttribute},
+			{"position", patch.Position, &ouser.Position, LS.PositionAttribute},
+			{"email", patch.Email, &ouser.Email, LS.EmailAttribute},
+		} {
+			if attr.patchValue != nil && *attr.patchValue != *attr.userValue && *attr.selector != "" {
+				return errors.New(attr.name)
+			}
+		}
+	} else if ouser.IsSAMLUser() {
+		for _, attr := range []struct {
+			name       string
+			patchValue *string
+			userValue  *string
+			selector   *string
+		}{
+			{"nickname", patch.Nickname, &ouser.Nickname, SS.NicknameAttribute},
+			{"position", patch.Position, &ouser.Position, SS.PositionAttribute},
+			{"email", patch.Email, &ouser.Email, SS.EmailAttribute},
+		} {
+			if attr.patchValue != nil && *attr.patchValue != *attr.userValue && *attr.selector != "" {
+				return errors.New(attr.name)
+			}
+		}
+	}
+	// If any login provider is used, then the username may not be changed
+	if ouser.AuthService != "" && patch.Username != nil && *patch.Username != ouser.Username {
+		return errors.New("username")
+	}
+	return nil
+}
+
```

Comment:
The flow of this function could be improved to make it more understandable. The current flow is checking basically in field order. 
Check Firstname/Lastname (then check Auth)
Check Nickname/Position/Email (if auth = saml, ldap)
Check Username (if auth !"")
Then each of these have additional checks to determine whether or not to process. (ie. ouser.IsLdap()). 

I would suggest rewriting this function to be based on the users authentication and check all fields need to be checked for that authentication. Then write a helper function that each "authentication" type will call with an array of fields to check. Remember for SAML and LDAP, the fields don't need checked if the attribute isn't set.